flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java

```diff
@@ -282,97 +303,232 @@ public Status process() throws EventDeliveryException {
     } catch (Throwable th) {
       // catch-all for any unhandled Throwable so that the transaction is
       // correctly rolled back.
-      if (transaction != null) {
-        try {
-          transaction.rollback();
-        } catch (Exception ex) {
-          LOG.error("Transaction rollback failed", ex);
-          throw Throwables.propagate(ex);
-        }
-      }
+      rollbackTransaction();
 
-      // close the writer and remove the its reference
-      writer.close();
+      // We don't want to close the writer because that will persist data
+      // that will be replayed. This will avoid duplicate data.
+      // This is also no different the the no-op close the writer does if
+      // it's already in the ERROR state. We can just set it to null and a
+      // new one will be created the next call to process.
       this.writer = null;
```

Comment:
I think this is only the case if we are not committing each batch. If any transaction has been committed for the current writer, then we must close it.